scripts/scraper-ng/rules/html-warn-unknown-headings.js

```diff
@@ -0,0 +1,46 @@
+const path = require("path");
+
+const { selectAll } = require("hast-util-select");
+
+const { loadRecipe } = require("./utils.js");
+
+const source = "html-warn-unknown-headings";
+
+/**
+ * Emit warnings when a page contains H2 headings that don't
+ * correspond to ingredients in the page's recipe.
+ */
+function warnUnknownHeadings() {
+  return function attacher(tree, file) {
+    const recipeName = path.basename(file.data.recipePath, ".yaml");
+    const recipe = loadRecipe(file.data.recipePath);
+    const expectedH2 = expectedH2Headings(recipe.body);
+    const actualH2 = selectAll(`h2`, tree);
+    for (const actual of actualH2) {
+      if (!expectedH2.includes(actual.properties.id)) {
+        file.message(
```

Comment:
I didn't do that in this PR because (1) the need seemed less clear here than in the ingredient handlers and (2) the current logger API is only really suitable for ingredient handlers.

I wanted the logger because (a) the previous situation led to lots of state being passed around, and (b) the logging code wanted a very consistent way of formatting IDs, which led to lots of duplicated code. And this looked like it would get worse as we added more ingredient handlers and especially as we organized ingredient handlers in reusable ways.

But in this case this doesn't seem to apply so much - the log call is just one line here, and we don't need to pass any extra state around.

We might anyway want to use the logger API, but as it stands the logger API is only suitable for ingredient handler errors (it expects to log the ingredient, which doesn't apply here, and also makes messages fatal, which doesn't apply). So we'd need to rework the API for this case. We might decide to to this at some point, but it didn't seem worth it in this PR.

Reply:
> We might decide to to this at some point, but it didn't seem worth it in this PR.

That sounds reasonable. I'll mark this resolved.